### DISCUSSION TITLE
Wisteria-Dataviz code cmake error about extra ‘;’

### DISCUSSION THREAD

**Discussion by chenliangabc** [Blake-Madden/Wisteria-Dataviz#48] [Q&A]
I get a lot of  errors  about extra ‘;’ when compiling code with CMake on Ubuntu.
Does it need any environment support, although I can solve it by removing ‘;’ .
![image](https://github.com/user-attachments/assets/81f5d7b5-f77c-4da6-8725-ce8529923257)

**Comment by Blake-Madden** [comment]
Yes, that ';' should be removed. I just checked in a fix for this, thanks!

**Reply by chenliangabc** [reply]
> Yes, that ';' should be removed. I just checked in a fix for this, thanks!

@Blake-Madden The problem with semicolons isn't limited to `points.h` files.
![image](https://github.com/user-attachments/assets/bbd80529-f777-48ef-acdc-f17e18a51672)

I'd like to take this opportunity to ask some other questions, thanks:
I want to update the data in the child thread and then refresh in the main thread, I don't know how to refresh, calling `Wisteria::Canvas::Refresh` doesn't work.

**Comment by Blake-Madden** [comment]
You will want to call `SetData()` for the particular graph after changing its data. You can call it the same way you did initially and that will force the graph to reload the data from it.

**Comment by Blake-Madden** [comment]
Just fixed the rest of the semicolons. Please let me know if I missed any.

**Reply by chenliangabc** [reply]
> Just fixed the rest of the semicolons. Please let me know if I missed any.

### There are also the following problems found(Ubuntu CMake)
1. CMakeLists.txt release mode, ```-Werror```
![image](https://github.com/user-attachments/assets/8bd2925c-35a9-4d64-b334-78a6b734775d)
2. All ```wxIMPLEMENT_DYNAMIC_CLASS```  have a semicolon problem.
![image](https://github.com/user-attachments/assets/bf2f6d5d-71ea-4e59-ba29-48e6388f8d1e)

### I want to ask a few questions, thanks
1. ```Dataset::Resize```, if I call ```AddCategoricalColumn(const wxString& columnName,constColumnWithStringTable::StringTableType& stringTable)``` before ```Resize```, ```stringTable``` is not empty. It Calls the code in the box, but it doesn't seem like it should be called.
![image](https://github.com/user-attachments/assets/cc6c8506-0ee1-4450-8b28-036fc98e6f36)
2. The following works, but it repeats a lot of steps. I changed the data and just want to repaint.
> You will want to call `SetData()` for the particular graph after changing its data. You can call it the same way you did initially and that will force the graph to reload the data from it.

**Reply by Blake-Madden** [reply]
I just checked in fixes for the first two points.

For the `Dataset::Resize` issue, are you saying that it is causing incorrect results, or that it is doing something that seems unnecessary?

For `SetData()`, this needs to be called for times when the graph needs to recalculate aggregates from the data (e.g., box plots) or other times when it needs to rebuild other internal values from the data.

**Reply by chenliangabc** [reply]
> For the `Dataset::Resize` issue, are you saying that it is causing incorrect results, or that it is doing something that seems unnecessary?

Yes, when call```SetData()```, Initially I passed the stringtable size and color size to be equal, so that the colors corresponded to each other.
In ```Resize```, it add one to the stringtable. After that, the resulting color is not a one-to-one correspondence.
![image](https://github.com/user-attachments/assets/e9e4f198-02b4-4cb2-abbc-a1e632612515)

#### The checkbox identifies the color I want
Incorrect results,  color is not a one-to-one correspondence.
![image](https://github.com/user-attachments/assets/7eca1fd2-9eeb-4133-8e2e-f602a69f55a1)

The correct result, by call ```SetStringTable``` after ```Resize```
![image](https://github.com/user-attachments/assets/84c462c7-96f5-475d-8212-6cb40e88b2ab)

**Reply by Blake-Madden** [reply]
I prefer `Resize()` verifying that categoricals have a missing data value in their string tables, and adding them if necessary; I see your point, however, that you don't want your string tables mutated. If you don't believe that you will have missing data, then this would be unnecessary and unexpected.

If I add an optional parameter to `Resize()` telling it to not perform the MD code check (and more importantly, not adding an MD code to the string table), would that work?

**Reply by chenliangabc** [reply]
> I prefer `Resize()` verifying that categoricals have a missing data value in their string tables, and adding them if necessary; I see your point, however, that you don't want your string tables mutated. If you don't believe that you will have missing data, then this would be unnecessary and unexpected.
> 
> If I add an optional parameter to `Resize()` telling it to not perform the MD code check (and more importantly, not adding an MD code to the string table), would that work?

Sorry for the late reply. You can do this, but I don't think it's necessary, just add some caveats.

Also, does it support chart translation and single-axis scaling?

**Comment by Blake-Madden** [comment]
No, scaling is done on both axes.

I'm not sure what chart translation is. Do you have an example?

**Reply by chenliangabc** [reply]
Translation error, is mouse move.
